File: lua/pe023.lua
-- this computes projecteuler.net problem 023

local MAX = 28123
local abundant = {}
local adictionary = {}

local function sum_factors(number)
  local total = 1
  local last = math.floor(math.sqrt(number))
  for i = 2, last do
    if (number % i) == 0 then
      local divisor = math.floor(number / i)
      total = i + total
      if divisor ~= i then
	total = divisor + total
      end
    end
  end
  return total 
end

local function table_contains_key(table, key)
    return table[key] ~= nil
end

local abundant_count = 0
for i = 1, MAX do
  local fsum = sum_factors(i)
  if i < fsum then
    abundant_count = abundant_count + 1
    abundant[abundant_count] = i
    adictionary[i] = fsum
  end
end

local asum = 0
for i = 1, abundant[1] - 1 do
  asum = i + asum
end

for i = abundant[1], MAX do
  local addi = true
  for j = 1, abundant_count do
    if abundant[j] < i then
      diff = i - abundant[j]
      if table_contains_key(adictionary, diff) then
        addi = false
        break
      end
    end
  end
  if addi then
    asum = i + asum
  end
end

print(asum)

File: python/pe023.py
#!/usr/bin/env python3

""" this computes projecteuler.net problem 023 """

import sys

from math import sqrt


def sum_factors(number):
    asum = 1
    for i in range(2, int(sqrt(number)) + 1):
        if ((number % i) == 0):
            divisor = number / i
            asum += i
            if (i != divisor):
                asum += divisor
    return asum


MAX = 28123
abundant = []
adictionary = {}
for i in range(1, MAX):
    fsum = sum_factors(i)
    if (i < fsum):
        abundant.append(i)
        adictionary[i] = fsum

asum = 0
for i in range(1, abundant[0]):
    asum += i
for i in range(abundant[0], MAX):
    addi = 1
    for j in abundant:
        if (j >= i):
            break
        diff = i - j
        if diff in adictionary:
            addi = 0
            break
    if (addi):
        asum += i

print(asum)
sys.exit(0)
